dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/ExecutionGraph.java

```diff
@@ -0,0 +1,318 @@
+            constraint = new EmptinessConstraint(innerGraph);
+        } else if (axiom.isIrreflexivity()) {
+            constraint = new IrreflexivityConstraint(innerGraph);
+        } else {
+            throw new UnsupportedOperationException("The axiom " + axiom + " is not recognized.");
+        }
+
+        constraintMap.put(axiom, constraint);
+        return constraint;
+    }
+
+    private EventGraph getGraphFromRelation(Relation rel) {
+        if (relationGraphMap.containsKey(rel)) {
+            return relationGraphMap.get(rel);
+        }
+
+        EventGraph graph;
+        Class<?> relClass = rel.getClass();
+
+        // ===== Filter special relations ======
+        if (SPECIAL_RELS.contains(rel.getName())) {
+            switch (rel.getName()) {
+                case CTRL:
+                    graph = new CtrlDepGraph();
+                    break;
+                case DATA:
+                    graph = new DataDepGraph();
+                    break;
+                case ADDR:
+                    graph = new AddrDepGraph();
+                    break;
+                case CRIT:
+                    graph = new RcuGraph();
+                    break;
+                default:
+                    throw new UnsupportedOperationException(rel.getName() + " is marked as special relation but has associated graph.");
+            }
+        } else if (relClass == RelRf.class) {
+            graph = rfGraph = new ReadFromGraph();
+        } else if (relClass == RelLoc.class) {
+            graph = locGraph = new LocationGraph();
+        } else if (relClass == RelPo.class) {
+            graph = poGraph = new ProgramOrderGraph();
+        } else if (relClass == RelCo.class) {
+            graph = coGraph = new TransitiveGraph(woGraph = new CoherenceGraph());
+            woGraph.setName("_wo");
+        } else if (rel.isRecursiveRelation()) {
+            RecursiveGraph recGraph = new RecursiveGraph();
+            recGraph.setName(rel.getName() + "_rec");
+            relationGraphMap.put(rel, recGraph);
+            recGraph.setConcreteGraph(getGraphFromRelation(rel.getInner()));
+            return recGraph;
+        } else if (rel.isUnaryRelation()) {
+            Relation innerRelation = rel.getInner();
+            EventGraph innerGraph = getGraphFromRelation(innerRelation);
+            // A safety check because recursion might compute this edge set
+            if (relationGraphMap.containsKey(rel)) {
+                return relationGraphMap.get(rel);
+            }
+
+            if (relClass == RelInverse.class) {
+                graph = new InverseGraph(innerGraph);
+            } else if (relClass == RelTrans.class) {
+                graph = new TransitiveGraph(innerGraph);
+            } else if (relClass == RelRangeIdentity.class) {
+                graph = new RangeIdentityGraph(innerGraph);
+            } else if (relClass == RelTransRef.class) {
+                RelTrans relTrans = new RelTrans(innerRelation);
```

Comment:
The first is not true because ```rel``` is of type ```RelTransRef``` and not ```RelTrans```.
It is the case that I should be able to avoid the newly created ```RelTrans``` and just create ```graph= new ReflexiveClosureGraph(new TransitiveGraph(innerGraph))```. However, this will cause an unfortunate discrepancy where I have an additional ```TransitiveGraph``` that has no associated ```Relation```. Ideally, we would have a ```ReflexiveClosureRel``` at the ```Wmm``` level and then I could directly match ```Relations``` and ```EventGraphs```.